src/systems/filecoin_vm/runtime/impl/runtime.go

```diff
@@ -308,6 +309,38 @@ func (rt *VMContext) _checkNumValidateCalls(x int) {
 	}
 }
 
+func (rt *VMContext) ValidateIsStorageMiner(minerAddr addr.Address) bool {
+	codeID, ok := rt.GetActorCodeID(minerAddr)
+	Assert(ok)
+	if !codeID.IsBuiltin() {
+		return false
+	}
+	return (codeID.As_Builtin() == actor.BuiltinActorID_StorageMiner)
+}
+
+func (rt *VMContext) ValidateImmediateCallerDetermineRecipient(entryAddr addr.Address) addr.Address {
+	if rt.ValidateIsStorageMiner(entryAddr) {
+		// Storage miner actor; implied funds recipient is the associated owner address.
+		ownerAddr, workerAddr := rt.GetMinerAccountsAssert(entryAddr)
+		rt.ValidateImmediateCallerInSet([]addr.Address{ownerAddr, workerAddr})
+		return ownerAddr
+	} else {
+		// Account actor (client); funds recipient is just the entry address itself.
+		rt.ValidateImmediateCallerAcceptAnyOfType(actor.BuiltinActorID_Account)
+		return entryAddr
+	}
+}
+
+func (rt *VMContext) GetMinerAccountsAssert(minerAddr addr.Address) (ownerAddr addr.Address, workerAddr addr.Address) {
+	ownerAddr = addr.Deserialize_Address_Compact_Assert(
+		rt.SendQuery(minerAddr, ai.Method_StorageMinerActor_GetOwnerAddr, []util.Serialization{}))
+
+	workerAddr = addr.Deserialize_Address_Compact_Assert(
+		rt.SendQuery(minerAddr, ai.Method_StorageMinerActor_GetWorkerAddr, []util.Serialization{}))
+
+	return
+}
```

Comment:
This should not be on the runtime! Do not couple the runtime to application specifics, and this can perfectly well be implemented outside the runtime.